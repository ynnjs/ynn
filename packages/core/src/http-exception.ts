/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: src/http-exception.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/04/2021
 * Description:
 ******************************************************************/

import statuses from 'statuses';
import { PartialKeys } from '@ynn/utility-types';

export interface HttpExceptionResponse {
    status: number;
    message: unknown;
    error: string;
}

export class HttpException extends Error {

    #status = 500;
    #message: string = statuses.message[ 500 ] ?? '';
    #response?: HttpExceptionResponse;

    constructor(
        status: number | string | Readonly<PartialKeys<HttpExceptionResponse, 'message' | 'error'>>,
        message?: string
    ) {
        super();
        if( typeof status === 'string' ) {
            this.message = status;
        } else if( typeof status === 'number' ) {
            this.status = status;
            message && ( this.message = message );
        } else {
            const response = { ...status };
            if( !response.error ) {
                response.error = statuses.message[ response.status ] ?? '';
            }
            if( !response.message ) response.message = '';
            this.response = response as HttpExceptionResponse;
        }
    }

    set status( status: number ) {
        this.#status = status;
        this.message = statuses.message[ status ] ?? '';
    }

    get status(): number {
        return this.#status;
    }

    set message( message: string ) {
        this.#message = message;
    }

    get message(): string {
        return this.#message;
    }

    get response(): HttpExceptionResponse | undefined {
        return this.#response;
    }

    set response( response: HttpExceptionResponse | undefined ) {
        this.#response = response;
        if( response ) {
            this.status = response.status;
            this.message = response.error;
        }
    }
}
