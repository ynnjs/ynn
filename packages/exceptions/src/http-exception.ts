/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: exceptions/http-exception.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 05/01/2021
 * Description:
 ******************************************************************/

import statuses from 'statuses';

export type HttpExceptionResponse = number
    | string
    | Record<string, any>; // eslint-disable-line @typescript-eslint/no-explicit-any

export class HttpException extends Error {

    static DEFAULT_STATUS = 500;

    #status: number = HttpException.DEFAULT_STATUS;
    #error: string = statuses.message[ HttpException.DEFAULT_STATUS ] ?? '';
    #response: HttpExceptionResponse;

    constructor( response: HttpExceptionResponse, error?: string ) {
        super();

        this.#response = response;

        if( typeof response === 'number' ) {
            this.status = response;
            error && ( this.error = error );
        } else if( typeof response === 'string' ) {
            this.error = response;
        } else {
            if( typeof response.status === 'number' ) {
                this.status = response.status;
            }

            if( typeof response.error === 'string' ) {
                this.error = response.error;
            }
        }
    }

    set status( status: number ) {
        this.#status = status;
        this.error = statuses.message[ status ] ?? '';
    }

    get status(): number {
        return this.#status;
    }

    set error( error: string ) {
        this.#error = error;
    }

    get error(): string {
        return this.#error;
    }

    get response(): HttpExceptionResponse {
        return this.#response;
    }

    toJSON(): HttpExceptionResponse {
        return {
            status : this.status,
            error : this.error,
            response : this.response
        };
    }
}
