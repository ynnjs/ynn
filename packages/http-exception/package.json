{
    "name": "@ynn/http-exception",
    "version": "0.0.0",
    "author": "LvChengbin <lvchengbin59@gmail.com>",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf lib",
        "build": "yarn clean && tsc -b -v",
        "test": "jest",
        "lint": "eslint {src,test}/**/*.ts --fix --cache"
    },
    "engines": {
        "node": ">14.15.0"
    },
    "main": "lib/index.js",
    "files": [
        "lib"
    ],
    "devDependencies": {
        "@types/statuses": "^2.0.0",
        "eslint": "^7.24.0",
        "eslint_d": "^10.0.4",
        "jest": "^26.6.3",
        "jest-extended": "^0.11.5",
        "ts-jest": "^26.5.4",
        "typescript": "^4.2.4"
    },
    "dependencies": {
        "@ynn/utility-types": "workspace:packages/utility-types",
        "statuses": "^2.0.1"
    }
}
