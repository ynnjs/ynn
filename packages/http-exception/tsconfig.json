{
    "extends" : "../tsconfig.base.json",
    "compilerOptions": {
        "outDir": "./lib/",
        "rootDir": "./src/",
        "composite": true,
        "tsBuildInfoFile": "./lib/tsconfig.tsbuildinfo"
    },
    "references" : [
        { "path" : "../utility-types" }
    ],
    "include" : [ "src/**/*" ]
}
