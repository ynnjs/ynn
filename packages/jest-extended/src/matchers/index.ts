/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: matchers/index.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/25/2021
 * Description:
 ******************************************************************/

import toThrowHttpException from './toThrowHttpException';

export default {
    toThrowHttpException
};
