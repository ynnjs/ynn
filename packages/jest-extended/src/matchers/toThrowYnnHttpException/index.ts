/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: toThrowYnnHttpException/index.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/25/2021
 * Description:
 ******************************************************************/

import { CustomMatcherResult } from 'jest';
import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';
import { VariadicFunction } from '@ynn/utility-types';
import { HttpException, HttpExceptionResponse } from '@ynn/http-exception';
import predicate from './predicate';

const passMessage = ( received: unknown, expected: unknown ): ( () => string ) => (): string => {
    return matcherHint( '.toThrowYnnHttpException', 'function', 'type' ) +
        '\n\n' +
        'Expected not to throw:\n' +
        `  ${printExpected( expected?.toJSON?.() ?? expected )}\n` +
        'Thrown:\n' +
        `  ${printReceived( received?.toJSON?.() ?? received )}\n`;
};

const failMessage = ( received: unknown, expected: unknown ): ( () => string ) => (): string => {
    return matcherHint( '.not.toThrowYnnHttpException', 'function', 'type' ) +
        '\n\n' +
        'Expected to throw:\n' +
        `  ${printExpected( expected?.toJSON?.() ?? expected )}\n` +
        'Thrown:\n' +
        `  ${printReceived( received?.toJSON?.() ?? received )}\n`;
};

/**
 *
 * @example
 * ```ts
 * expect( () => { throw new HttpException( 400 ) } ).toThrowYnnHttpException()
 * ```
 *
 * @param callback - the callback function which will throw exception
 * @param response - status, message, response or HttpException instance
 * @param message - error message of HttpException instance.
 */
export default function toThrowYnnHttpException(
    callback: VariadicFunction,
    response?: number | string | HttpExceptionResponse | HttpException,
    error?: string
): CustomMatcherResult {

    console.log( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx', callback );

    if( !callback || typeof callback !== 'function' ) {
        return {
            pass : false,
            message : (): string => `Recevied value must be a function but instead "${callback}" was found`
        };
    }

    let exception: any; // eslint-disable-line @typescript-eslint/no-explicit-any

    try { callback() } catch( e: unknown ) { exception = e }

    if( !exception ) {
        return {
            pass : false,
            message : (): string => 'Expect the fucntion to throw an HttpException.\n But it didn\'t throw anything'
        };
    }

    if( !( exception instanceof HttpException ) ) {
        return {
            pass : false,
            // message : failMessage( error, HttpException )
            message : (): string => `Expect the function to throw an HttpException.\n But ${exception} was thrown`
        };
    }

    /**
     * any HttpException instance will pass the checking if the response argument is empty
     */
    if( !response ) {
        return {
            pass : true,
            message : passMessage( exception, HttpException )
        };
    }

    if( typeof response === 'number' ) {
        if( error === undefined ) {
            if( exception.status === response ) {
                return {
                    pass : true,
                    message : passMessage( exception, { status : response } )
                };
            }

            return {
                pass : false,
                message : failMessage( exception, { status : response } )
            };
        }
        if( exception.status === response && exception.error === error ) {
            return {
                pass : true,
                message : passMessage( exception, { status : response, error } )
            };
        }

        return {
            pass : false,
            message : failMessage( exception, { status : response, error } )
        };
    }

    if( typeof response === 'string' ) {
        if( exception.error === response ) {
            return {
                pass : true,
                message : passMessage( exception, { error : response } )
            };
        }
        return {
            pass : false,
            message : failMessage( exception, { error : response } )
        };
    }

    let httpException: HttpException;

    if( response instanceof HttpException ) {
        httpException = response;
    } else {
        httpException = new HttpException( response, error );
    }

    if( !predicate( exception, response, error ) ) {
        return {
            pass : false,
            message : failMessage( exception, httpException )
        };
    }

    return {
        pass : true,
        message : passMessage( exception, new HttpException( response, error ) )
    };
}
