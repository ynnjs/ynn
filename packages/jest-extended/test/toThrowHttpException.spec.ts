/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: test/toThrowHttpException.spec.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/25/2021
 * Description:
 ******************************************************************/

import { HttpException } from '@ynn/http-exception';
import matchers from '../src/matchers';

expect.extend( matchers );

describe( '.toThrowHttpException', () => {
    it( 'passes when thrown an arbitrary HttpException', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException();
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException();
        expect( () => { throw new HttpException( { status : 400, message : [ 'Something Error' ] } ) } ).toThrowHttpException();
    } );

    it( 'passes when thrown an HttpException matched the expect HttpException instance', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException( new HttpException( 400 ) );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( new HttpException( 400, 'Something Error' ) );
        const response = { status : 400, message : [ 'Something Error' ] };
        expect( () => { throw new HttpException( response ) } ).toThrowHttpException( new HttpException( response ) );
    } );

    it( 'passes when thrown an HttpException and matched the given status', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException( 400 );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( 400 );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( 400 );
        expect( () => { throw new HttpException( { status : 400, message : [ 'Something Error' ] } ) } ).toThrowHttpException( 400 );
    } );

    it( 'passes when thrown an HttpException and matched the given error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException( 'Bad Request' );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( 'Something Error' );
        expect( () => { throw new HttpException( { status : 400, error : 'Something Error' } ) } ).toThrowHttpException( 'Something Error' );
    } );

    it( 'passes when thrown an HttpException and matched both the given status and error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException( 400, 'Bad Request' );
        expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( 400, 'Something Error' );
        expect( () => { throw new HttpException( { status : 400, error : 'Something Error' } ) } ).toThrowHttpException( 400, 'Something Error' );
    } );

    it( 'fails when not thrown any exception', () => {
        expect( () => {
            expect( () => {} ).toThrowHttpException();
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when not thrown HttpException', () => {
        expect( () => {
            expect( () => { throw new Error() } ).toThrowHttpException();
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown an HttpException but not matched the given HttpException instance', () => {
        expect( () => {
            expect( () => { throw new HttpException( 401 ) } ).toThrowHttpException( new HttpException( 400 ) );
        } ).toThrowErrorMatchingSnapshot();
        expect( () => {
            expect( () => { throw new HttpException( 400, 'Something Error' ) } ).toThrowHttpException( new HttpException( 400 ) );
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown an HttpException and not matched the given status', () => {
        expect( () => {
            expect( () => { throw new HttpException( 401 ) } ).toThrowHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( 401, 'something error' ) } ).toThrowHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( { status : 401, error : 'something error' } ) } ).toThrowHttpException( 400 );
        } ).toThrowErrorMatchingSnapshot();
    } );

    it( 'fails when thrown an HttpException and not matched the given error message', () => {
        expect( () => {
            expect( () => { throw new HttpException( 400 ) } ).toThrowHttpException( 'Something Error' );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( 400, 'something error' ) } ).toThrowHttpException( 400, 'Bad Request' );
        } ).toThrowErrorMatchingSnapshot();

        expect( () => {
            expect( () => { throw new HttpException( { status : 401, error : 'something error' } ) } ).toThrowHttpException( 400, 'Bad Request' );
        } ).toThrowErrorMatchingSnapshot();
    } );
} );

describe( '.not.toThrowHttpException', () => {
    it( 'passes when not thrown any exception', () => {
        expect( () => {} ).not.toThrowHttpException();
    } );

    it( 'passes when not thrown a HttpException', () => {
        expect( () => { throw new Error() } ).not.toThrowHttpException();
    } );

    it( 'passes when thrown a HttpException but not matched the given status', () => {
        expect( () => { throw new HttpException( 400 ) } ).not.toThrowHttpException( 401 );
    } );

    it( 'passes when thrown a HttpException but not matched the given error message', () => {
        expect( () => { throw new HttpException( 400 ) } ).not.toThrowHttpException( 'Something Error' );
    } );

    it( 'passes when thrown a HttpException but not matched the given response', () => {
        expect( () => { throw new HttpException( { status : 400, message : [ 'Bad Request' ] } ) } ).not.toThrowHttpException( new HttpException( {
            status : 400,
            message : [ 'Something Error' ]
        } ) );
    } );
} );
