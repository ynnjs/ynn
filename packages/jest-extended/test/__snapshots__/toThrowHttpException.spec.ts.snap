// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.toThrowHttpException fails when not thrown HttpException 1`] = `
"Expect the function to throw an HttpException.
 But Error was thrown"
`;

exports[`.toThrowHttpException fails when not thrown any exception 1`] = `
"Expect the fucntion to throw an HttpException.
 But it didn't throw anything"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given error message 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Something Error\\"}[39m
Thrown:
  [31m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given error message 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given error message 3`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": {\\"error\\": \\"something error\\", \\"status\\": 401}, \\"status\\": 401}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given status 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Unauthorized\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given status 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException and not matched the given status 3`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"something error\\", \\"response\\": {\\"error\\": \\"something error\\", \\"status\\": 401}, \\"status\\": 401}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException but not matched the given HttpException instance 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Unauthorized\\", \\"response\\": 401, \\"status\\": 401}[39m
"
`;

exports[`.toThrowHttpException fails when thrown an HttpException but not matched the given HttpException instance 2`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowHttpException([22m[32mtype[39m[2m)[22m

Expected to throw:
  [32m{\\"error\\": \\"Bad Request\\", \\"response\\": 400, \\"status\\": 400}[39m
Thrown:
  [31m{\\"error\\": \\"Something Error\\", \\"response\\": 400, \\"status\\": 400}[39m
"
`;
